src/main/java/graphql/execution/instrumentation/dataloader/FieldLevelTrackingApproach.java

```diff
@@ -160,6 +179,31 @@ public void onFieldValuesException() {
         };
     }
 
+    private int getCurrentLevel(InstrumentationExecutionStrategyParameters parameters) {
+        ResultPath path = parameters.getExecutionStrategyParameters().getPath();
+        return path.getLevel() + 1;
+    }
+
+    @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+    private static void increaseCallCounts(CallStack callStack, int curLevel, InstrumentationExecutionStrategyParameters parameters) {
+        int fieldCount = parameters.getExecutionStrategyParameters().getFields().size();
+        synchronized (callStack) {
+            callStack.increaseExpectedFetchCount(curLevel, fieldCount);
+            callStack.increaseHappenedStrategyCalls(curLevel);
+        }
+    }
+
+    @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+    private void onFieldValuesInfoDispatchIfNeeded(CallStack callStack, List<FieldValueInfo> fieldValueInfoList, int curLevel) {
+        boolean dispatchNeeded;
+        synchronized (callStack) {
+            dispatchNeeded = handleOnFieldValuesInfo(fieldValueInfoList, callStack, curLevel);
+        }
+        if (dispatchNeeded) {
+            dispatch();
+        }
+    }
+
```

Comment:
Move the code down here to be common.  No other change other than beginObjectResolution has been introduced